src/components/tl/cuda/tl_cuda_cache.c

```diff
@@ -0,0 +1,326 @@
+#include "core/ucc_team.h"
+#include <cuda_runtime.h>
+
+#define ENABLE_CACHE 1
+
+static ucs_pgt_dir_t*
+ucc_tl_cuda_cache_pgt_dir_alloc(const ucs_pgtable_t *pgtable)
+{
+    void *ptr;
+    int ret;
+
+    ret = posix_memalign(&ptr, ucs_max(sizeof(void *), UCS_PGT_ENTRY_MIN_ALIGN),
+                             sizeof(ucs_pgt_dir_t));
+    return (ret == 0) ? ptr : NULL;
+}
+
+static void ucc_tl_cuda_cache_pgt_dir_release(const ucs_pgtable_t *pgtable,
+                                              ucs_pgt_dir_t *dir)
+{
+    free(dir);
+}
+
+
+static void
+ucc_tl_cuda_cache_region_collect_callback(const ucs_pgtable_t *pgtable,
+                                           ucs_pgt_region_t *pgt_region,
+                                           void *arg)
+{
+    ucs_list_link_t *list = arg;
+    ucc_tl_cuda_cache_region_t *region;
+
+    region = ucs_derived_of(pgt_region, ucc_tl_cuda_cache_region_t);
+    ucs_list_add_tail(list, &region->list);
+}
+
+static void ucc_tl_cuda_cache_purge(ucc_tl_cuda_cache_t *cache)
+{
+    ucc_tl_cuda_cache_region_t *region, *tmp;
+    ucs_list_link_t region_list;
+
+    ucs_list_head_init(&region_list);
+    ucs_pgtable_purge(&cache->pgtable, ucc_tl_cuda_cache_region_collect_callback,
+                      &region_list);
+    ucs_list_for_each_safe(region, tmp, &region_list, list) {
+        CUDACHECK_NORET(cudaIpcCloseMemHandle(region->mapped_addr));
+        free(region);
+    }
+}
+
+ucc_status_t ucc_tl_cuda_create_cache(ucc_tl_cuda_cache_t **cache,
+                                      const char *name)
+{
+
+    ucc_status_t status;
+    ucc_tl_cuda_cache_t *cache_desc;
+    int ret;
+
+    cache_desc = ucc_malloc(sizeof(ucc_tl_cuda_cache_t), "ucc_tl_cuda_cache_t");
+    if (cache_desc == NULL) {
+        ucs_error("failed to allocate memory for tl_cuda cache");
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    ret = pthread_rwlock_init(&cache_desc->lock, NULL);
+    if (ret) {
+        ucs_error("pthread_rwlock_init() failed: %m");
+        status = UCC_ERR_INVALID_PARAM;
+        goto err;
+    }
+
+    if (UCS_OK != ucs_pgtable_init(&cache_desc->pgtable,
+                                   ucc_tl_cuda_cache_pgt_dir_alloc,
+                                   ucc_tl_cuda_cache_pgt_dir_release)) {
+        goto err_destroy_rwlock;
+    }
+
+    cache_desc->name = strdup(name);
+    if (cache_desc->name == NULL) {
+        status = UCC_ERR_NO_MEMORY;
+        goto err_destroy_rwlock;
+    }
+
+    *cache = cache_desc;
+    return UCC_OK;
+
+err_destroy_rwlock:
+    pthread_rwlock_destroy(&cache_desc->lock);
+err:
+    free(cache_desc);
+    return status;
+
+    return UCC_OK;
+}
+
+
+void ucc_tl_cuda_destroy_cache(ucc_tl_cuda_cache_t *cache)
+{
+    ucc_tl_cuda_cache_purge(cache);
+    ucs_pgtable_cleanup(&cache->pgtable);
+    pthread_rwlock_destroy(&cache->lock);
+    free(cache->name);
+    free(cache);
+}
+
+#if ENABLE_CACHE
+static void ucc_tl_cuda_cache_invalidate_regions(ucc_tl_cuda_cache_t *cache,
```

Comment:
@Sergei-Lebedev @bureddy can this cache stuff be abstracted as generic peer_based cache? I see a use case for KNEM regions cache in TL/SHM ?